### ISSUE TITLE
Pickling objects with __slots__ fails

### ISSUE THREAD

**Issue by sk1p** [cloudpipe/cloudpickle#225]
The following minimal test case:

```diff
diff --git a/tests/cloudpickle_test.py b/tests/cloudpickle_test.py
index d318da7..fec779c 100644
--- a/tests/cloudpickle_test.py
+++ b/tests/cloudpickle_test.py
@@ -1310,6 +1310,17 @@ class CloudPickleTest(unittest.TestCase):
 
         self.assertEqual(f2.__annotations__, f.__annotations__)
 
+    def test_slots(self):
+        class WithSlots(object):
+            __slots__ = ["a"]
+
+            def __init__(self):
+                self.a = 42
+
+        obj = WithSlots()
+        obj2 = pickle_depickle(obj, protocol=self.protocol)
+        self.assertEqual(obj2.a, 42)
+
 
 class Protocol2CloudPickleTest(CloudPickleTest):

```

Fails with this exception:

```
obj = <tests.cloudpickle_test.CloudPickleTest.test_slots.<locals>.WithSlots object at 0x7f5cdc8f94c8>, protocol = 4

    def pickle_depickle(obj, protocol=cloudpickle.DEFAULT_PROTOCOL):
        """Helper function to test whether object pickled with cloudpickle can be
        depickled with pickle
        """
>       return pickle.loads(cloudpickle.dumps(obj, protocol=protocol))
E       AttributeError: type object 'WithSlots' has no attribute 'a'

obj        = <tests.cloudpickle_test.CloudPickleTest.test_slots.<locals>.WithSlots object at 0x7f5cdc8f94c8>
protocol   = 4
```

Pickling with `pickle` from the standard library works.

**Comment by pwais** [comment]
Similarly, `pickle` can pickle a slotted object with highest protocol, but `cloudpickle` cannot:

```
# python
Python 2.7.12 (default, Dec  4 2017, 14:50:18) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> class WithSlots(object):
...   __slots__ = ["a"]
...   def __init__(self):
...     self.a = 42
... 
>>> import pickle
>>> pickle.dumps(WithSlots())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/pickle.py", line 1380, in dumps
    Pickler(file, protocol).dump(obj)
  File "/usr/lib/python2.7/pickle.py", line 224, in dump
    self.save(obj)
  File "/usr/lib/python2.7/pickle.py", line 306, in save
    rv = reduce(self.proto)
  File "/usr/lib/python2.7/copy_reg.py", line 77, in _reduce_ex
    raise TypeError("a class that defines __slots__ without "
TypeError: a class that defines __slots__ without defining __getstate__ cannot be pickled
>>> pickle.dumps(WithSlots(), -1)
'\x80\x02c__main__\nWithSlots\nq\x00)\x81q\x01N}q\x02U\x01aq\x03K*s\x86q\x04b.'

>>> import cloudpickle
>>> cloudpickle.__version__
'0.6.1'
>>> cloudpickle.dumps(WithSlots())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/cloudpickle/cloudpickle.py", line 931, in dumps
    cp.dump(obj)
  File "/usr/local/lib/python2.7/dist-packages/cloudpickle/cloudpickle.py", line 284, in dump
    return Pickler.dump(self, obj)
  File "/usr/lib/python2.7/pickle.py", line 224, in dump
    self.save(obj)
  File "/usr/lib/python2.7/pickle.py", line 331, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python2.7/pickle.py", line 396, in save_reduce
    save(cls)
  File "/usr/lib/python2.7/pickle.py", line 286, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/local/lib/python2.7/dist-packages/cloudpickle/cloudpickle.py", line 678, in save_global
    return self.save_dynamic_class(obj)
  File "/usr/local/lib/python2.7/dist-packages/cloudpickle/cloudpickle.py", line 517, in save_dynamic_class
    save(clsdict)
  File "/usr/lib/python2.7/pickle.py", line 286, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python2.7/pickle.py", line 655, in save_dict
    self._batch_setitems(obj.iteritems())
  File "/usr/lib/python2.7/pickle.py", line 687, in _batch_setitems
    save(v)
  File "/usr/lib/python2.7/pickle.py", line 306, in save
    rv = reduce(self.proto)
TypeError: can't pickle member_descriptor objects
```

**Comment by pwais** [comment]
Ah! this only occurs when the object is in the `__main__` module.  It also happens for older cloudpickle (I tried 0.3.1) so I don't think it's a regression.  

This works:
```
$ cat foo.py 
class Bar(object):
  __slots__ = ('x',)
  def __init__(self):
    self.x = 5

$ python
Python 2.7.12 (default, Dec  4 2017, 14:50:18) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from foo import Bar
>>> import cloudpickle
>>> cloudpickle.dumps(Bar())
'\x80\x02cfoo\nBar\nq\x00)\x81q\x01N}q\x02U\x01xq\x03K\x05s\x86q\x04b.'
```

My guess is the test and/or demo don't work due to this `__main__` module issue.

**Comment by ogrisel** [comment]
Thanks for the report @sk1p and @pwais. Is any of you interested in working on a pull request to implement support for slots in cloudpickle?

@pierreglaser you might be interested in this as well.

**Comment by pierreglaser** [comment]
Indeed this should be implemented. I'll work on this.

**Comment by ogrisel** [comment]
@pwais @sk1p I just released cloudpickle 0.7.0 to PyPI for this new feature.
